### PR TITLE
A practical demo example - Sizzle Restaurant

### PR DESCRIPTION
**Pull request by ChloePlanet**
This example shows:
- ~human
- ~ui
- ~db
- package
- long description
- sequence of ...
- alias
- empty project
- types/tables/endpoints
- import OpenAPI3.0 file

TODO:
- convert endpoints to RESTful endpoints
- add database operation

### CODE REVIEW

File: demo/sizzle.sysl [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+import mastercard.yaml as MasterCard
+
+# This app specifies the projects to include, it is optional and can be left out
+Sizzle[~project]:
+    @contact.name = "Jimmy Smith"
+    @description =:
+        | Sizzle is a Gourmet Sausage Restaurant.
+        |
+        | We aim to provide the truly Aussie sausage sizzle 
+        | to our customers.
```
Comment (cuminandpaprika) [inline demo/sizzle.sysl]:
```suggestion
        | We aim to provide an authentic Aussie sausage sizzle experience
        | to our customers.
```

File: demo/sizzle.sysl [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    @version = "1.0.0"
+
+    Frontend:
+        Application
+    Backend:
+        BFF
+        Database
+
+
+Dine-in Customer [~human]:
```
Comment (cuminandpaprika) [inline demo/sizzle.sysl:46]: Love the modelling of the customers, this is a great example for documenting use-cases!

Comment (cuminandpaprika) [reply demo/sizzle.sysl:46]: @orlade-anz might be interested to see this also

File: demo/sizzle.sysl [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+        last_name <: string
+        phone <: string
+        email <: string
+        password <: string
+
+    !type UnauthorizedError:
+        error_msg <: string
+
+ProductServer:
+    @package="BFF"
```
Comment (cuminandpaprika) [inline demo/sizzle.sysl]:
```suggestion
    @package="ProductServer"
```

File: demo/sizzle.sysl [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+
+
+
+Common:
+    !type Empty[~empty]:
+        @description="Empty Type"
+        ...
+
+IdentityServer:
+    @package="BFF"
```
Comment (cuminandpaprika) [inline demo/sizzle.sysl]:
```suggestion
    @package="IdentityServer"
```

File: demo/sizzle.sysl [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+        @description="Product information"
+        id <: int
+        name <: string
+        image <: string
+        details <: string
+        price <: int
+
+
+OrderServer:
+    @package="BFF"
```
Comment (cuminandpaprika) [inline demo/sizzle.sysl]:
```suggestion
    @package="OrderServer"
```

File: demo/sizzle.sysl [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+
+    !enum OrderStatus:
+        created: 1
+        placed: 2
+        shipped: 3
+        delivered: 4
+    
+
+PaymentServer:
+    @package="BFF"
```
Comment (cuminandpaprika) [inline demo/sizzle.sysl]:
```suggestion
    @package="PaymentServer"
```

File: demo/sizzle.sysl [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+    @description =: 
+        | This server handles all the payment related endpoints.
+
+    Pay:
+        if processor_type == "visa":
+            Visa <- Pay
+        else if processor_type == "mastercard":
+            MasterCard <- POST /pay
```
Comment (cuminandpaprika) [inline demo/sizzle.sysl:259]: Seems like the diagram isn't showing the call to MasterCard, which is interesting
image

File: demo/sizzle.sysl [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+        cash: 1
+        card: 2
+
+Visa[~external]:
+    # ~external specifies an external service
+    Pay:
+        ...
+
+DeliveryServer:
+    @package="BFF"
```
Comment (cuminandpaprika) [inline demo/sizzle.sysl]:
```suggestion
    @package="DeliveryServer"
```

File: demo/sizzle.sysl [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+    @description =: 
+        | We are going to provide delivery service ASAP
```
Comment (cuminandpaprika) [inline demo/sizzle.sysl:277]: 😂

File: demo/sizzle.sysl [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+        | since our customers need it during COVID-19
+    ...
+
+
+
+MegaDatabase[~db]:
+    # ~db specifies a database
```
Comment (cuminandpaprika) [inline demo/sizzle.sysl:284]: This generates a really nice type diagram!

image

File: demo/sizzle.sysl [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+        card_number <: string
+        expiry <: string
+
+    !table reviews:
+        id <: int [~pk]
+        score <: int
+        comment <: string
+
+Cache:
+    @pacakge="Database"
```
Comment (cuminandpaprika) [inline demo/sizzle.sysl]:
```suggestion
    @package="Database"
```

File: demo/sizzle.sysl [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+    CustomerProfile(customer_id <: int):
+        # TODO: database interaction
+        return ok <: Customer
+
+    UpdatePassword(customer_id <: int, old <: string, new <: string):
+        # DB update
+        return ok
+
+    !type Customer:
+        @description="Empty Type"
```
Comment (cuminandpaprika) [inline demo/sizzle.sysl]:
```suggestion
        @description="This contains all information relating to a customer"
```

File: demo/sizzle.sysl [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+    Place and Pay Order:
+        Website <- PlaceOrder
+        Website <- Pay
+    Review:
+        Website <- Review
+
+
+Mobile [~ui]:
+    # ~ui specifies a user interface
+    @package = "Application"
```
Comment (cuminandpaprika) [inline demo/sizzle.sysl:80]:
```suggestion
    @package = "iOS"
```

File: demo/sizzle.sysl [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+    Order:
+        Website <- Order
+    Place and Pay Order:
+        Website <- PlaceOrder
+        Website <- Pay
+    Review:
+        Website <- Review
+
+
+Mobile [~ui]:
```
Comment (cuminandpaprika) [inline demo/sizzle.sysl:78]: Maybe we can duplicate this so that we have both iOSApp and AndroidApp?

**Review by cuminandpaprika — APPROVED** [review]
Looks amazing, a few nitpicks

File: demo/sizzle.sysl (hunk 7 above)
Comment (ChloePlanet) [reply demo/sizzle.sysl:259]: Yes that is a new issue. Also this integration diagram should show "Dine-in Customer -> Dashboard -> PaymentServer" but it doesn't.
image

File: demo/sizzle.sysl (hunk 10 above)
Comment (ChloePlanet) [reply demo/sizzle.sysl:284]: Yeah I like it

File: demo/sizzle.sysl (hunk 14 above)
Comment (ChloePlanet) [reply demo/sizzle.sysl:78]: I wanna keep this simple at first. And in an ideal world, the iOSApp and AndroidApp are supposed to provide the same functionality.

File: demo/sizzle.sysl (hunk 10 above)
Comment (orlade-anz) [reply demo/sizzle.sysl:284]: It looks great! The data model itself is a bit odd though:

- Each review should probably have a customer as an author, not be related to an order.
- I think card_payment is just a card?
- All names should be consistently singular or plural
- What is orders_products.comments?
- payment_details.type is a float?

Nits:
- Probably get rid of customers.middle_name?
- Maybe add some dates to things - dates are always a pain

Comment (ChloePlanet) [reply demo/sizzle.sysl:284]: orders_products.comments is something like "less chilli", "more sugar", etc.

**Pull request merged**